energy-dao/src/external_sc_interactions/farm_interactions.rs

```diff
@@ -0,0 +1,252 @@
+        if user_rewards.amount > 0 {
+            farm_config_mapper.update(|config| config.reward_reserve -= &user_rewards.amount);
+        }
+        let new_attributes = WrappedFarmTokenAttributes {
+            token_rps: farm_config.farm_rps,
+        };
+        self.burn_tokens(
+            &payment.token_identifier,
+            payment.token_nonce,
+            &payment.amount,
+        );
+        let new_farm_token = self.mint_tokens(
+            farm_config.wrapped_token_id,
+            payment.amount,
+            &new_attributes,
+        );
+
+        let mut user_payments = ManagedVec::new();
+        user_payments.push(new_farm_token);
+        if user_rewards.amount > 0 {
+            let wrapper_user_rewards = self.wrap_locked_token(user_rewards);
+            user_payments.push(wrapper_user_rewards);
+        }
+        let caller = self.blockchain().get_caller();
+        self.send().direct_multi(&caller, &user_payments);
+    }
+
+    #[payable("*")]
+    #[endpoint(unstakeFarm)]
+    fn unstake_farm(&self, farm_address: ManagedAddress) -> EsdtTokenPayment {
+        let payment = self.call_value().single_esdt();
+        let mut farm_config_mapper = self.farm_config(&farm_address);
+        require!(!farm_config_mapper.is_empty(), ERROR_FARM_DOES_NOT_EXIST);
+        let farm_config = farm_config_mapper.get();
+        require!(
+            farm_config.wrapped_token_id == payment.token_identifier,
+            ERROR_BAD_PAYMENT_TOKENS
+        );
+
+        let current_farm_position = EsdtTokenPayment::new(
+            farm_config.farm_token_id.clone(),
+            farm_config.farm_token_nonce,
+            farm_config.farm_staked_value.clone(),
+        );
+        let claim_rewards_result = self.call_farm_claim(farm_address, current_farm_position);
+        let (new_farm_token, farm_rewards) = claim_rewards_result.into_tuple();
+        self.update_farm_after_claim(
+            &farm_config,
+            &mut farm_config_mapper,
+            new_farm_token.clone(),
+            farm_rewards,
+        );
+
+        let user_rewards = self.compute_user_rewards_payment(&mut farm_config_mapper, &payment);
+
+        farm_config_mapper.update(|config| {
+            config.farm_staked_value -= &payment.amount;
+            config.farm_unstaked_value += &payment.amount;
+            if user_rewards.amount > 0 {
+                config.reward_reserve -= &user_rewards.amount
+            }
+        });
+
+        self.burn_tokens(
+            &payment.token_identifier,
+            payment.token_nonce,
+            &payment.amount,
+        );
+        let current_epoch = self.blockchain().get_block_epoch();
+        let unstake_attributes = UnstakeTokenAttributes {
+            unstake_epoch: current_epoch,
+            token_nonce: new_farm_token.token_nonce,
+        };
+        let unstake_token_payment = self.mint_tokens(
+            farm_config.unstake_token_id.clone(),
```

Comment:
Don't think you need the clone here.
```suggestion
            farm_config.unstake_token_id,
```

Reply:
Updated.